stateproof/builder.go

```diff
@@ -20,48 +20,167 @@ import (
 	"context"
 	"database/sql"
 	"encoding/binary"
+	"errors"
 	"fmt"
 	"sort"
 
 	"github.com/algorand/go-algorand/config"
 	"github.com/algorand/go-algorand/crypto/stateproof"
 	"github.com/algorand/go-algorand/data/basics"
-	"github.com/algorand/go-algorand/data/bookkeeping"
 	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
 	"github.com/algorand/go-algorand/logging"
 	"github.com/algorand/go-algorand/network"
 	"github.com/algorand/go-algorand/protocol"
 	"github.com/algorand/go-algorand/stateproof/verify"
 )
 
-// makeBuilderForRound not threadsafe, should be called in a lock environment
-func (spw *Worker) makeBuilderForRound(rnd basics.Round) (builder, error) {
-	l := spw.ledger
-	hdr, err := l.BlockHdr(rnd)
+var errVotersNotTracked = errors.New("voters not tracked for the given lookback round")
+
+// OnPrepareVoterCommit is a function called by the voters tracker when it's preparing to commit rnd. It gives the builder
+// the chance to persist the data it needs.
+func (spw *Worker) OnPrepareVoterCommit(oldBase basics.Round, newBase basics.Round, votersFetcher ledgercore.LedgerForSPBuilder) {
+	for rnd := oldBase + 1; rnd <= newBase; rnd++ {
+		header, err := votersFetcher.BlockHdr(rnd)
+		if err != nil {
+			spw.log.Errorf("OnPrepareVoterCommit(%d): could not fetch round header: %v", rnd, err)
+			continue
+		}
+
+		proto := config.Consensus[header.CurrentProtocol]
+		if proto.StateProofInterval == 0 || uint64(rnd)%proto.StateProofInterval != 0 {
+			continue
+		}
+
+		var builderExists bool
+		err = spw.db.Atomic(func(ctx context.Context, tx *sql.Tx) (err error) {
+			builderExists, err = builderExistInDB(tx, rnd)
+			return err
+		})
+		if err != nil {
+			spw.log.Warnf("OnPrepareVoterCommit(%d): could not check builder existence, assuming it doesn't exist: %v\n", rnd, err)
+		} else if builderExists {
+			continue
+		}
+
+		buildr, err := createBuilder(rnd, votersFetcher)
+		if err != nil {
+			if errors.Is(err, errVotersNotTracked) {
+				// Voters not tracked for that round.  Might not be a valid
+				// state proof round; state proofs might not be enabled; etc.
+				spw.log.Warnf("OnPrepareVoterCommit(%d): %v", rnd, err)
+				continue
+			}
+
+			spw.log.Errorf("OnPrepareVoterCommit(%d): could not create builder: %v", rnd, err)
+			continue
+		}
+
+		err = spw.db.Atomic(func(_ context.Context, tx *sql.Tx) error {
+			return persistBuilder(tx, rnd, &buildr)
```

Comment:
This call can come at the same time when `initBuilders` or `handleSig` is being called. In that case, one of them will error due to primary key violation.
Is this possible?
If so, perhaps it is better to de-register `OnPrepareVoterCommit` listener when stopping the worker and register again (already done) when starting?

Reply:
as part of our last change, the handler is being registered only after `initBuilders` is being called (in a synced manner),   so `handleSig ` will not happen

Reply:
How about `handleSig`?

Reply:
`handleSig` might occur only after  `initBuilders` is completed.

Reply:
handleSig conflicts with OnPrepareVoterCommit, not initBuilders

Reply:
possible resolution for (handleSig conflicts with OnPrepareVoterCommit)
1. persistBuilder does not err if the same builder (rnd, data) are inserted. Rationale is a builder created from commits and sigs must be the same.
2. write test

Reply:
there is proposed query prepared query
```golang
encoded = protocol.Encode(b)
tx.Exec("INSERT INTO builders(rnd, builder)
SELECT ?, ? 
WHERE NOT EXISTS(SELECT 1 FROM builders WHERE rnd = ? AND text = ?)",
rnd, encoded, rnd, encoded)
```